File: src/leetcode141.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

struct ListNode {
    int val;
    struct ListNode *next;
};
//#双指针，#差分器，#链表
bool hasCycle(struct ListNode *head) {
	if (head==NULL || head->next==NULL)
	{
		return false;
	}
	
    struct ListNode *fast=head->next,*slow=head;
	while (fast!=slow)
	{
		if (fast==NULL||slow==NULL)
		{
			return false;
		}
		
		fast=fast->next->next;
		slow=slow->next;

	}
	return true;
}

File: src/leetcode141.py
# hash 表，链表
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
""" 环检测，直接hash表 """
class Solution:
	def hasCycle(self, head: Optional[ListNode]) -> bool:
		a=[]
		while(head!=None and head.next!=None ):
			if id(head) in a:
				return True
			else:
				a.append(id(head))
				head=head.next
		
		return False

if __name__ == "__main__":
	s=Solution()
	s.hasCycle(None)